unix_integration/resolver/src/idprovider/system.rs

```diff
@@ -0,0 +1,126 @@
+use hashbrown::HashMap;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+use super::interface::{Id, IdpError};
+use kanidm_unix_common::unix_passwd::{EtcGroup, EtcUser};
+use kanidm_unix_common::unix_proto::{NssGroup, NssUser};
+
+pub struct SystemProviderInternal {
+    users: HashMap<Id, Arc<EtcUser>>,
+    user_list: Vec<Arc<EtcUser>>,
+    groups: HashMap<Id, Arc<EtcGroup>>,
+    group_list: Vec<Arc<EtcGroup>>,
+}
+
+pub struct SystemProvider {
+    inner: Mutex<SystemProviderInternal>,
+}
+
+impl SystemProvider {
+    pub fn new() -> Result<Self, IdpError> {
+        Ok(SystemProvider {
+            inner: Mutex::new(SystemProviderInternal {
+                users: Default::default(),
+                user_list: Default::default(),
+                groups: Default::default(),
+                group_list: Default::default(),
+            }),
+        })
+    }
+
+    pub async fn reload(&self, users: Vec<EtcUser>, groups: Vec<EtcGroup>) {
+        let mut system_ids_txn = self.inner.lock().await;
+        system_ids_txn.users.clear();
+        system_ids_txn.user_list.clear();
+        system_ids_txn.groups.clear();
+        system_ids_txn.group_list.clear();
+
+        for group in groups {
+            let name = Id::Name(group.name.clone());
+            let gid = Id::Gid(group.gid);
+            let group = Arc::new(group);
+
+            if system_ids_txn.groups.insert(name, group.clone()).is_some() {
+                error!(name = %group.name, gid = %group.gid, "group name conflict");
+            };
+            if system_ids_txn.groups.insert(gid, group.clone()).is_some() {
+                error!(name = %group.name, gid = %group.gid, "group id conflict");
+            }
+            system_ids_txn.group_list.push(group);
+        }
+
+        for user in users {
+            let name = Id::Name(user.name.clone());
+            let uid = Id::Gid(user.uid);
+            let gid = Id::Gid(user.gid);
+
+            if user.uid != user.gid {
+                error!(name = %user.name, uid = %user.uid, gid = %user.gid, "user uid and gid are not the same, this may be a security risk!");
+            }
+
+            // Security checks.
+            if let Some(group) = system_ids_txn.groups.get(&gid) {
+                if group.name != user.name {
+                    error!(name = %user.name, uid = %user.uid, gid = %user.gid, "user private group does not appear to have the same name as the user, this may be a security risk!");
+                }
+                if !(group.members.is_empty()
+                    || (group.members.len() == 1 && group.members.first() == Some(&user.name)))
+                {
+                    error!(name = %user.name, uid = %user.uid, gid = %user.gid, "user private group must not have members, THIS IS A SECURITY RISK!");
+                }
+            } else {
+                warn!(name = %user.name, uid = %user.uid, gid = %user.gid, "user private group is not present on system, synthesising it");
```

Comment:
since this is (likely) pretty common, this feels like a debug-level message?

Reply:
I kind of want there to be a stick that says "your distro is doing dumb shit" but the other problem yeh, is how to force people to fix it. nfi.